### PR TITLE
Replace EigenBroadcast with ElementwiseBroadcast in ReduceGrad

### PR DESCRIPTION
**Pull request by AnnaTrainingG**
### PR types

Performance optimization 
### PR changes

OPs
### Describe

Reduce EigenBroadcastcase with ElementwiseBroadcast in ReduceGrad
为扩大KP算子覆盖率，统一将Reduce_sum/mean 反向的Eigen适配代码替换为ElementwiseBroadcast Kernel
ReduceGrad 性能统计对比：

op name | case | axise | dtype | 优化前 （us) | 优化后 （us) | speed   up
-- | -- | -- | -- | -- | -- | --
reduce_sum_grad | [-1L, 2048L, 33L, 33L] | [2,3] | float32 | 626 | 160.73 | 3.89
reduce_sum_grad | [-1L, 2048L, 33L, 33L] | [2,3] | float16 | 633 | 83.64 | 7.57
reduce_sum_grad | [-1L, 8L, 128L] | [1] | float32 | 2.44 | 1.71 | 1.43
reduce_sum_grad | [-1L, 8L, 128L] | [1] | float16 | 2.84 | 1.7 | 1.67
reduce_sum_grad | [30522L, 1024L] | [] | float32 | 150.38 | 138.79 | 1.08
reduce_sum_grad | [30522L, 1024L] | [] | float16 | 89.81 | 43.44 | 2.07
op name | case | axise | dtype | 优化前 （us) | 优化后 （us) | speed up
reduce_mean_grad | [-1L, 2048L, 33L, 33L] | [2,3] | float32 | 672 | 160.71 | 4.18
reduce_mean_grad | [-1L, 2048L, 33L, 33L] | [2,3] | float16 | 681 | 83.1 | 8.19
reduce_mean_grad | [-1L, 8L, 128L] | [1] | float32 | 3.115 | 1.72 | 1.81
reduce_mean_grad | [-1L, 8L, 128L] | [1] | float16 | 3.171 | 1.66 | 1.91
reduce_mean_grad | [30522L, 1024L] | [] | float32 | 152.78 | 138.83 | 1.10
reduce_mean_grad | [30522L, 1024L] | [] | float16 | 134.96 | 69.835 | 1.93

benchmark 异常说明，与本次PR修改无关，本地测试无性能影响：

op   name | shape | dtype | dev (us) | new (us) | speed up
-- | -- | -- | -- | -- | --
p_norm_1 backward | [300, 128, 128], axis = -1 porder = 3.0 | float32 | 178.26 | 178.06 | 1.00112322
matmul_3_backward | [paddle][p_norm] p_norm {       run_tf: True       run_torch: True       axis: -1       porder: 3.0       keepdim: False       x_shape: [300, 128, 128]       x_dtype: float32       atol: 1e-06     [paddle][p_norm] p_norm {       run_tf: True       run_torch: True       axis: -1       porder: 3.0       keepdim: False       x_shape: [300, 128, 128]       x_dtype: float32       atol: 1e-06 | float32 | 12.7867 | 12.865 | 0.99391372
matmul_9_forward | [paddle][matmul] matmul {       run_tf: True       run_torch: True       atol: 1.0       transpose_x: False       transpose_y: False       x_shape: [4, 12, 64, 85]       x_dtype: float16       y_shape: [4, 12, 85, 512]       y_dtype: float16     } | float32 | 26.165 | 26.146 | 1.00072669

### CODE REVIEW

File: paddle/fluid/operators/reduce_ops/reduce_op.h [hunk 1]
```diff
@@ -391,7 +392,7 @@ template <typename DeviceContext, typename T, typename Functor,
           bool kNoNeedBufferX = false, bool kNoNeedBufferY = false>
 class ReduceGradKernel : public framework::OpKernel<T> {
  public:
-  void ComputeFromInput(const Tensor* input2,
+  void ComputeFromInput(const framework::Tensor* input2,
```
Comment (ZzSean) [inline paddle/fluid/operators/reduce_ops/reduce_op.h]: 这里之前不加`framework::`不会报错？

File: paddle/fluid/operators/reduce_ops/reduce_op.h [hunk 2]
```diff
@@ -724,6 +725,37 @@ class ReduceCudaKernel : public framework::OpKernel<T> {
...
   }
 };
+
+template <typename T, template <typename, typename> class TransformOp>
+class ReduceCudaGradKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    bool reduce_all = context.Attr<bool>("reduce_all");
+    std::vector<int> dims = context.Attr<std::vector<int>>("dim");
+    auto* input0 = context.Input<Tensor>("X");
```
Comment (ZzSean) [inline paddle/fluid/operators/reduce_ops/reduce_op.h]: 这个input0的命名跟其他的tensor好像不太匹配？

File: paddle/fluid/operators/reduce_ops/reduce_op.h (hunk 1 above)
Comment (AnnaTrainingG) [reply paddle/fluid/operators/reduce_ops/reduce_op.h]: 之前没报错

File: paddle/fluid/operators/reduce_ops/reduce_op.h (hunk 2 above)
Comment (AnnaTrainingG) [reply paddle/fluid/operators/reduce_ops/reduce_op.h]: 已经修改

**Review by xingfeng01 — DISMISSED** [review]

File: paddle/pten/kernels/gpu/elementwise.h [hunk 3]
```diff
@@ -414,12 +414,20 @@ struct DimensionsTransform {
...
+      // when ins.size() = 1 set N = in.size() + 1
+      in_dims.resize(N);
+      in_dims[0] = paddle::framework::vectorize<int64_t>(ins[0]->dims());
+      in_dims[1] = out_dims;
+      // Add out_dims to in_dims to avoid errors in dims merging
+    } else {
+      in_dims.resize(N);
+      for (int j = 0; j < N; ++j) {
+        in_dims[j] = paddle::framework::vectorize<int64_t>(ins[j]->dims());
+      }
```
Comment (JamesLim-sy) [inline paddle/pten/kernels/gpu/elementwise.h:147]: `DimensionsTransform`设计之初过多考虑了二元及以上的计算进行广播操作，有些忽视针对一元计算的部分，这里的修改方式能通通过校验，但是修改显得较为晦涩不够直观。根据代码判断问题点出在了L460-L475部分，这里应该是加`N==1` 的判断语句，来控制L460-L475部分是否执行，可以尝试一下这种修改方式，应该也能够通过编译和正确性校验
```
if (N != 1) {
    int min_idx = 0;
    int min_val = std::accumulate(
        in_dims[0].begin(), in_dims[0].end(), 1, std::multiplies<int64_t>());
    for (int j = 1; j < N; ++j) {
      int temp = std::accumulate(
          in_dims[j].begin(), in_dims[j].end(), 1, std::multiplies<int64_t>());
      min_val = min_val > temp ? temp : min_val;
      min_idx = min_val == temp ? j : min_idx;
    }
    std::swap(in_dims[0], in_dims[min_idx]);
    // To Merge the dimension of input_tensors while the consequtive
    // 1-value-dimensions appears.
    merge_ptr = merge_sequential_one_dims;
    MergeDimensions<MergeFunctor>(merge_ptr, N);
    std::swap(in_dims[min_idx], in_dims[0]);
}
```

Comment (AnnaTrainingG) [reply paddle/pten/kernels/gpu/elementwise.h:147]: 后续PR 再进行修改

**Review by zhiqiu — APPROVED** [review]
LGTM for shareDataWith

**Review by xingfeng01 — APPROVED** [review]

**Review by ZzSean — APPROVED** [review]
LGTM for ci op benchmark

**Pull request merged**